### ISSUE TITLE
Some questions about the models

### ISSUE THREAD

**Issue by 14H034160212** [nuric/deeplogic#3]
Hi,
I have a question for the model. LSTM, MAC, DMN, IMA, IMASM, IMARSM, FWIMARSM. Does the full name of IMASM mean IMA + softmax, IMARSM mean IMA + rule + softmax, and FWIMARSM mean forward chaining + IMA + rule + softmax? So except the FWIMARSM, other models are all backward chaining? Isn't it?

Is another question have you public the σ model? Cause I see the σ in table 3 from the paper but do not find the related code.

And what is the main difference between multi-task learning and curriculum learning from the paper and code? Especially for the IMA model. I see there are both multi-task learning and curriculum learning version for the IMA model. But from the code, I did not see any difference from those two versions.

Besides, for the visualization of the predicate embedding, how do you draw those figures? Like figure 3 - figure 5.

Hope not disturbing you.
All the best

**Comment by nuric** [comment]
Hello,

Yes you got the naming scheme correct. It wouldn't be correct to say backward chaining as that is an paradigm learnt by model. The difference between FWIMARSM is that it processes the body of the rule and then the head, so it has a bias towards performing forward chaining whereas the other models start from the head and process body literals.

The sigma model stands for the sigmoid version, so plain `ima.py` is the sigmoid version while `imasm.py` is the softmax version.

Multi-task is when all the tasks are in a single dataset with models run for 4 iterations. Whereas with curriculum learning, we train on tasks 1 2 with 1 iteration, then reload that model and continue training on tasks 1 2 3 7 9 12 with 2 iterations and so on. So the models learns to perform 1 step deduction before 2 step tasks etc. The code allows you to you pick the tasks and then the iterations, which tasks and which iterations are used are presented in the paper.

The visualisations are done inside `eval.py`, Figure 3 is from `plot_single_preds()` and Figure 5 is computed from `plot_struct_preds()` and `plot_rules()` respectively.

Hope that helps, if you are not reproducing results but trying to run as baselines etc. I would recommend potentially using latest APIs to benefit from improvements and major changes done to the libraries since the experiments were run, especially for Keras and Tensorflow.
